### DISCUSSION TITLE
Is matching EOF considered a "consume" for the repeat combinator?

### DISCUSSION THREAD

**Discussion by DavidAntliff** [winnow-rs/winnow#523] [Q&A] [answered]
Hi,

My experience with winnow is limited to a couple of days so perhaps I've missed something, or there's a better way to do what I want to do.

I wish to match on a "blank line", and to be clear I mean:

 * The input is text,
 * A blank line starts at either the beginning of the input, or immediately after a `line_ending`,
 * A blank line has zero or more white-space characters,
 * A blank line ends in either a `line_ending` or an `eof`.

The problem I'm having is that even though this seems fairly simple:

```rust
enum Expression {
    BlankLine,
    // ... other types of lines
}

fn blank_line(i: &mut &str) -> PResult<Expression> {
    (
        space0,
        alt((line_ending, eof)),
    )
        .map(|_| Expression::BlankLine)
        .parse_next(i)
}
```

this does not seem to work with the `repeat` combinator:

```rust
fn expression(i: &mut &str) -> PResult<Expression> {
    alt((
        // ... other parsers for other types of lines
        blank_line,
    ))
    .parse_next(i)
}

fn file(i: &mut &str) -> PResult<Vec<Expression>> {
    repeat(0.., expression).parse_next(i)
}
```

```
`repeat` parsers must always consume
```

I think this is because the `eof` match is not considered a consumable part of the input, so there's a condition (zero whitespace, followed by EOF) that should match but doesn't consume anything.

I'm also not sure if this error is due a check performed when this condition arises (if ever), or if the check is done based on what _might_ happen by examining the parser tree. Is there a workaround?

**Comment by epage** [comment]
Yes, `eof` does not consume, ie it does not move what `input` points to 

One way to workaround this would be
```rust
enum Expression {
    BlankLine,
    // ... other types of lines
}

fn blank_line(i: &mut &str) -> PResult<Expression> {
    alt((
        (space0, line_ending),
        (space1, eof),
    ))
        .map(|_| Expression::BlankLine)
        .parse_next(i)
}
```
This could be optimized as
```rust
enum Expression {
    BlankLine,
    // ... other types of lines
}

fn blank_line(i: &mut &str) -> PResult<Expression> {
    alt((
        line_ending,
        (space1, alt((line_ending, eof))),
    ))
        .map(|_| Expression::BlankLine)
        .parse_next(i)
}
```
There might be additional optimizations that could be made with `dispatch!`

**Reply by DavidAntliff** [reply]
Thanks for the reply. I've studied this and I've realised my case is slightly more complex than my first example suggested. I think the problem I have is that I may have to add this kind of `alt((line_ending, (SomeSymbol, alt((line_ending, eof)))` pattern to every expression parser. Let me see if I can demonstrate:

```rust
enum Expression {
    A,  // in reality A & B are much more complex
    B,
    BlankLine,
}

fn a(i: &mut &str) -> PResult<Expression> {
    "a".parse_next(i)?;
    Ok(Expression::A)
}

fn b(i: &mut &str) -> PResult<Expression> {
    "b".parse_next(i)?;
    Ok(Expression::B)
}

fn expression(i: &mut &str) -> PResult<Expression> {
    alt((
        a,
        b,
        blank_line,
    ))
    .parse_next(i)
}

struct Program {
    expressions: Vec<Expression>,
}

fn program(i: &mut &str) -> PResult<Program> {
    // Handling EOF:
    //   EOF can occur at the end of a line (after the final line_ending), or after at least one
    //   valid symbol after a line_ending.
    //   Thus, a line will end in either a line_ending, or EOF, but if it ends in EOF, it must
    //   have at least one valid symbol prior. This could be white-space, an A, or a B, but ideally
    //   each expression parser shouldn't have to worry about checking for EOF.

    let expressions = repeat(
        0..,
        delimited(
            space0,
            expression,
            (space0, alt((line_ending, eof))),
            // I need to find a way to refactor to make the (space0, BlankLine, space0, eof) input not possible,
            // but if the input ends at anywhere other than a line_ending, then the symbol(s) before the eof could be *anything* valid
        ),
    )
    .parse_next(i)?;
    Ok(Program { expressions })
}
```

The problem I seem to have is that implementing your workaround means that every expression (including `BlankLine`) would need to have an `alt()` combinator for testing for possible EOF. Please see the large comment block in `fn program()`, which I hope explains why this is so.

Beyond `BlankLine`, the case where you have `(space1, alt((line_ending, eof)))`, `space1` would have to be "at least 1" of any valid symbol (as well as `space1`), which means repeating the entire sub-parser in this branch. Something like this I think:

```rust
    let expressions = repeat(
        0..,
        alt((
            delimited(space0, alt((a, b)), alt((line_ending, eof))),
            terminated(alt((space1, alt((a, b, blank_line))), alt((line_ending, eof)))
        ),
    )
    .parse_next(i)?;
```

I'm still trying to figure out exactly what this would look like, as my real code is more complex than these examples, so the `alt()` combinators get quite cumbersome.

Makes me wonder if I'm handling blank lines incorrectly - I need to record their existence to preserve line numbers, but maybe I need to refactor the parsers to move the `alt((line_ending, eoff))` into each expressions' parser?

Related, I was wondering if the `repeat` loop detection is done at the time the parser is instantiated, or when it actually encounters a loop? Because, if it's done early (the first time `repeat` is encountered), `eof` could be specially handled in the chain to be considered "consumable" (or "loop breaking") and avoid this kind of workaround, or if it's detected at runtime while parsing, then a quick look-ahead check against EOF before raising the error might also work.

----

EDIT: for context, here are some examples that the parser should accept:

Ends with a newline (emits A, BlankLine, B):
```
a

b
<EOF>
```

Ends without a trailing newline (emits A, BlankLine, B):

```
a

b<EOF>
```

A single blank line with trailing newline (emits BlankLine):

```

<EOF>
```

Ends with a blank line of non-zero length and no trailing newline (emits A, B, BlankLine):

```
a
b
     <EOF>
```

**Reply by DavidAntliff** [reply]
BTW, in case it helps anyone else, the code given above:

```rust
fn blank_line(i: &mut &str) -> PResult<Expression> {
    alt((
        line_ending,
        (space1, alt((line_ending, eof))),
    ))
        .map(|_| Expression::BlankLine)
        .parse_next(i)
}
```

Does not compile:
```
error[E0271]: type mismatch resolving `<&str as Stream>::Slice == (&str, &str)`
   --> src/asm/parser.rs:218:10
    |
218 |     alt((line_ending, (space1, alt((line_ending, eof)))))
    |     ---  ^^^^^^^^^^^ expected `(&str, &str)`, found `&str`
    |     |
    |     required by a bound introduced by this call
    |
    = note:  expected tuple `(&str, &str)`
            found reference `&str`
```

This is caused by mapping to different result types between `line_ending` and the 2-tuple, such that the following does compile by reducing the 2-tuple to a 1-tuple:

```rust
fn blank_line(i: &mut &str) -> PResult<Expression> {
    alt((line_ending, terminated(space1, alt((line_ending, eof)))))
        .map(|_| Expression::BlankLine)
        .parse_next(i)
}
```
or ignoring both match results entirely:
```rust
fn blank_line(i: &mut &str) -> PResult<Expression> {
    alt((line_ending.void(), (space1, alt((line_ending, eof))).void()))
        .map(|_| Expression::BlankLine)
        .parse_next(i)
}
```

**Reply by DavidAntliff** [reply]
Unfortunately I just couldn't get this to work. I ended up having to match for `alt((line_ending, eof))` in many of my sub-parsers, rather than in just one place, which meant I also had to start testing for trailing "comments" everywhere too, including those with a preceding `space0` and those without. It became a huge mess, sadly. I'll have another crack another time. Maybe an explicit "per line" loop at the top level, parsing each line individually, would work better than `repeat(0.., ...)`, so I can avoid the overly-eager loop check entirely.

Everything else in `winnow` has been very nice to use - nice work, thank you :)

**Reply by epage** [reply]
Do you have your grammar in BNF form?  That would make it easier to understand for me to provide suggestions

**Reply by DavidAntliff** [reply]
> Do you have your grammar in BNF form? That would make it easier to understand for me to provide suggestions

I think this sums up my situation, with some notes in comments:

```EBNF
<File> ::= <LineNewLine>* <LineNoNewLine>?

<Line> ::= (<A> | <B> | <BlankLine>) <WhiteSpace>? <Comment>?
<LineNewLine> ::= <Line> <NewLine>
<LineNoNewLine> ::= <Line>

/* in reality, there are a LOT of these "line" rules, A, B, C, ... Z */
<A> ::= "aaa"
<B> ::= "bbb"

<WhiteSpace> ::= (" " | "\t")*
<BlankLine> ::= <WhiteSpace>
<NewLine> ::= ("\n" | "\r\n")

/* simplified - in reality accept all characters to end of line/file */
<Comment> ::= ";" ([a-z] | [0-9] | " ")*
```

Some valid inputs:

Trailing newline:
```
aaa
<EOF>
```

No trailing newline:
```
aaa<EOF>
```

Trailing whitespace and newline:

```
aaa
    <EOF>
```

Trailing comment:
```
aaa
bbb  ; comment<EOF>
```

Blank lines:
```
aaa

bbb

<EOF>
```

[BNF Playground](http://bnfplayground.pauliankline.com/?bnf=%3CFile%3E%20%3A%3A%3D%20%3CLineNewLine%3E*%20%3CLineNoNewLine%3E%3F%0A%0A%3CLine%3E%20%3A%3A%3D%20(%3CA%3E%20%7C%20%3CB%3E%20%7C%20%3CBlankLine%3E)%20%3CWhiteSpace%3E%3F%20%3CComment%3E%3F%0A%3CLineNewLine%3E%20%3A%3A%3D%20%3CLine%3E%20%3CNewLine%3E%0A%3CLineNoNewLine%3E%20%3A%3A%3D%20%3CLine%3E%0A%0A%2F*%20in%20reality%2C%20there%20are%20a%20LOT%20of%20these%20%22line%22%20rules%2C%20A%2C%20B%2C%20C%2C%20...%20Z%20*%2F%0A%3CA%3E%20%3A%3A%3D%20%22aaa%22%0A%3CB%3E%20%3A%3A%3D%20%22bbb%22%0A%0A%3CWhiteSpace%3E%20%3A%3A%3D%20(%22%20%22%20%7C%20%22%5Ct%22)*%0A%3CBlankLine%3E%20%3A%3A%3D%20%3CWhiteSpace%3E%0A%3CNewLine%3E%20%3A%3A%3D%20(%22%5Cn%22%20%7C%20%22%5Cr%5Cn%22)%0A%0A%2F*%20simplified%20-%20in%20reality%20accept%20all%20characters%20to%20end%20of%20line%2Ffile%20*%2F%0A%3CComment%3E%20%3A%3A%3D%20%22%3B%22%20(%5Ba-z%5D%20%7C%20%5B0-9%5D%20%7C%20%22%20%22)*%0A&name=)

Putting this together has actually shed some light on the subject - I can perhaps better see how to compose a winnow parser from this, without mentioning `eof` at all, potentially. I will give it a go.

**Reply by epage** [reply]
There are also subtle grammar changes you can make that might simplify things.

For example, you have
```ebnf
<File> ::= <LineNewLine>* <LineNoNewLine>?
```
TOML has
```ebnf
toml = expression *( newline expression )
```

Also, isn't
```ebnf
<Line> ::= (<A> | <B> | <BlankLine>) <WhiteSpace>? <Comment>?
```
just
```ebnf
<Line> ::= (<A> | <B>)? <WhiteSpace>? <Comment>?
```

Also, since `WhiteSpace` is
```ebnf
<WhiteSpace> ::= (" " | "\t")*
```
then you can drop `?`
```ebnf
<Line> ::= (<A> | <B>)? <WhiteSpace> <Comment>?
```

However, I would have `WhiteSpace` defined as
```ebnf
<WhiteSpace> ::= (" " | "\t")
```
and leave the repetition up to the caller.

In case its useful, TOML's abnf is at https://github.com/toml-lang/toml/blob/main/toml.abnf

`toml_edit` is a Winnow parser that documents how each parser function maps back to the ABNF.

**Comment by Nahor** [comment, marked as answer]
`eof` doesn't not consume because it can't, by definition. It means "end-of-file", i.e "there-is-nothing-left-to-consume".

As for your problem, you have "repeat(line_w_newline or line_wo_newline)", which implies that you could have "\<line_wo_newline\>\<line_w_newline\>", which makes no sense. So what you need is "repeat(line_w_newline) then maybe line_wo_newline" (as you said yourself in your BNF grammar)

But probably a better way of thinking is to not see the newline character as part of the line, but as a separator (see `separated()`)

**Reply by DavidAntliff** [reply]
Hi @Nahor, thank you for the suggestion. Switching `repeat(0.., delimited(space0, ..., (space0, line_ending)))` to `separated(0.., delimited(space0, ..., space0), line_ending)` does the trick in my case. It sometimes adds an extra `BlankLine` to the end of the resultant vector but I think that's perfectly consistent with the input, and easily dealt with.

I'm not sure why `repeat(0..` hits the infinite-loop check but `separated(0..` does not - is that because the separated check always assumes it will consume the separator, which in my case is `line_ending` thus never "nothing"?

**Reply by Nahor** [reply]
>I'm not sure why `repeat(0..` hits the infinite-loop check

Because `delimited(space0, ..., (space0, line_ending))` can match a 0-length string ("\<zero space\>\<blankline terminating because of eof\>\<zero space\>"), i.e. it can always match, even when there are no more characters left.

`separated(...)` works because, as you figured, it requires newlines to continue matching. So once it's at EOF and there is no more newlines, it can't continue.

`delimited(...)` should work if you follow your own BNF:
```
(repeat(0.., line_newline), opt(line_no_newline))
aka 
(repeat(0.., (line, line_ending)), line)
```
where, like `separated`, you repeat while you still have line endings, then have one extra `line` to match the last line, even if it's just an empty/0-length blankline.

> It sometimes adds an extra BlankLine to the end of the resultant vector but I think that's perfectly consistent with the input, and easily dealt with.

I assume that it's when your document terminates with a newline character, in which case, there is an empty/0-length blankline between the last newline and eof.

**Reply by DavidAntliff** [reply]
> I assume that it's when your document terminates with a newline character, in which case, there is an empty/0-length blankline between the last newline and eof.

Yep, that's the one. Makes sense and is consistent.

Empty input "" also parses to `vec![BlankLine]` but that's consistent too.

Thanks!
